tables/validate.py

```diff
@@ -0,0 +1,132 @@
+        "not_nullable": "The cell is blank but is required.",
+        "field_uniqueness": "You entered duplicate data. Remove or replace the duplicate data.",
+        "multiple_fields_uniqueness": "You entered duplicate data. Remove or replace the duplicate data.",
+    }
+    IGNORED_FAILURES = {
+        "failure_case": r"TypeError\(.*",  # catches check failures on data of the wrong type
+        "check": r"dtype\(.*",  # catches dtype failures
+    }
+
+    def __init__(self, schema_config: dict[str, Any]) -> None:
+        """Initialises a TableValidator.
+
+        :param schema_config: A dictionary containing schema configuration.
+            The schema configuration should include information about columns, data types,
+            and any additional validation rules.
+        :raises ValueError: If the schema has coercion enabled.
+        """
+        if schema_config.get("coerce") or any(column.coerce for column in schema_config.get("columns", {}).values()):
+            raise ValueError("Coercion not supported.")
+        self.schema = pa.DataFrameSchema(**schema_config)
+
+    def validate(self, table: Table) -> None:
+        """Validates a table against a schema.
+
+        :param table: table to validate
+        :raises ValueError: If the table's columns do not match the schema.
+        :raises TableValidationError: If the table is invalid.
+        """
+        self._check_columns(table)
+        try:
+            self.schema.validate(table.df, lazy=True)
+        except pa.errors.SchemaErrors as schema_errors:
+            failures = self._get_failures(schema_errors)
+            failures = self._filter_ignored(failures)
+            if failures:
+                validation_errors = [self._parse_failure(failure, table) for failure in failures]
+                raise TableValidationError(validation_errors=validation_errors)
+
+    def _check_columns(self, table: Table):
+        if cols_in_df_not_in_schema := set(table.df.columns).difference(set(self.schema.columns.keys())):
+            raise ValueError(f"Table columns {cols_in_df_not_in_schema} are not in the schema.")
+        if cols_in_schema_not_in_df := set(self.schema.columns.keys()).difference(set(table.df.columns)):
+            raise ValueError(f"Schema columns {cols_in_schema_not_in_df} are not in the table.")
+
+    @staticmethod
+    def _get_failures(schema_errors: pa.errors.SchemaErrors) -> list[Failure]:
+        return [Failure(**failure._asdict()) for failure in schema_errors.failure_cases.itertuples(index=False)]
+
+    def _parse_failure(self, failure: Failure, table: Table) -> ValidationError:
+        excel_cell_ref = (
+            table.get_excel_cell(failure.index, failure.column).excel_ref
+            if pd.notnull(failure.index) and pd.notnull(failure.column)
+            else None
+        )
+        message = self.MESSAGE_OVERRIDE.get(failure.check, failure.check)
+        return ValidationError(excel_cell_ref=excel_cell_ref, message=message)
```

Comment:
Minor thing but the kwargs are passed out of order with the definition in `ValidationError` - makes no difference in practice as they are kwargs, but for the sake of style worth re-ordering